packages/myst-toc/src/toc.ts

```diff
@@ -1,21 +1,193 @@
-import type { TOC } from './types.js';
-import schema from './schema.json';
-import _Ajv from 'ajv';
+import type { TOC, Entry, FileEntry, URLEntry, PatternEntry, ParentEntry } from './types.js';
+
+import type { ValidationOptions } from 'simple-validators';
+import {
+  defined,
+  incrementOptions,
+  validateBoolean,
+  validateEnum,
+  validateList,
+  validateNumber,
+  validateObjectKeys,
+  validateString,
+  validationError,
+} from 'simple-validators';
+
+const COMMON_ENTRY_KEYS = ['title', 'hidden', 'numbering', 'id', 'part', 'class'];
+
+function validateCommonEntry(entry: any, opts: ValidationOptions): Entry | undefined {
+  if (defined(entry.title)) {
+    entry.title = validateString(entry.title, incrementOptions('title', opts));
+  }
+
+  if (defined(entry.hidden)) {
+    entry.hidden = validateBoolean(entry.hidden, incrementOptions('hidden', opts));
+  }
+
+  if (defined(entry.numbering)) {
+    entry.numbering = validateString(entry.numbering, incrementOptions('numbering', opts));
+  }
+
+  if (defined(entry.id)) {
+    entry.id = validateString(entry.id, incrementOptions('id', opts));
+  }
+
+  if (defined(entry.part)) {
+    entry.part = validateString(entry.part, incrementOptions('part', opts));
+  }
+
+  if (defined(entry.class)) {
+    entry.class = validateString(entry.class, incrementOptions('class', opts));
+  }
+
+  return entry as Entry;
+}
+
+export function validateFileEntry(entry: any, opts: ValidationOptions): FileEntry | undefined {
+  let outputEntry = validateObjectKeys(
+    entry,
+    {
+      required: ['file'],
+      optional: [...COMMON_ENTRY_KEYS, 'children'],
+    },
+    opts,
+  );
+  if (!outputEntry) {
+    return undefined;
+  }
+
+  outputEntry.file = validateString(outputEntry.file, incrementOptions('file', opts));
+
```

Comment:
```suggestion
if (!outputEntry.file) return undefined
```

We should have this here - if `file` is defined but invalid we want to short-circuit, not return ann invalid `FileEntry`.

Reply:
Usually how I do this, so types are a little more helpful, is have an intermediate object:

- input `entry: any`
- then, like `const entryAsObject: Record<string, any>` as the result of `validateObjectKeys`
- finally `const outputEntry: FileEntry` as the final output

With that, we no longer have to cast `as FileEntry` at the end of the function, and instead get nice type checks throughout. Hopefully that makes sense...

Something like this: https://github.com/executablebooks/mystmd/blob/main/packages/myst-config/src/site/validators.ts#L51-L62 - which has `input`, intermediate `value`, and final, typed `output`.